segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/ServiceStarter.java

```diff
@@ -191,15 +192,14 @@ private void attachZKSegmentManager(ServiceBuilder builder) {
 
     private CuratorFramework createZKClient() {
         if (this.serviceConfig.isSecureZK()) {
-            System.setProperty(SECURE_CLIENT, Boolean.toString(this.serviceConfig.isSecureZK()));
-            System.setProperty(ZOOKEEPER_CLIENT_CNXN_SOCKET, "org.apache.zookeeper.ClientCnxnSocketNetty");
-            System.setProperty(SSL_TRUSTSTORE_LOCATION, this.serviceConfig.getZkTrustStore());
-            System.setProperty(SSL_TRUSTSTORE_PASSWD, JKSHelper.loadPasswordFrom(this.serviceConfig.getZkTrustStorePasswordPath()));
+            ZKTLSUtils.setSecureZKClientProperties(this.serviceConfig.getZkTrustStore(),
+                    JKSHelper.loadPasswordFrom(this.serviceConfig.getZkTrustStorePasswordPath()));
         }
         CuratorFramework zkClient = CuratorFrameworkFactory
                 .builder()
                 .connectString(this.serviceConfig.getZkURL())
                 .namespace("pravega/" + this.serviceConfig.getClusterName())
+                .zookeeperFactory(new ZKClientFactory())
```

Comment:
Observing that the curator 4.0.1 uses `DefaultZookeeperFactory` which is similar to `ZKClientFactory `.  The code snippets links are : 
https://github.com/apache/curator/blob/apache-curator-4.0.1/curator-framework/src/main/java/org/apache/curator/framework/CuratorFrameworkFactory.java#L66
and 
https://github.com/apache/curator/blob/apache-curator-4.0.1/curator-client/src/main/java/org/apache/curator/utils/DefaultZookeeperFactory.java#L27

Do let me know if there is a different reason for having this class...

Reply:
If you see the code for `ZKClientFactory`, when we invoke `newZookeeper` for the first time, the `client == null` and then the object internally keeps the info passed to create the Zookeeper client. This includes the `connectString`, that is the `hostname` of the zookeeper server to connect. Subsequent calls to `newZookeeper` use such info store in the `ZKClientFactory`, not the one passed by parameter. This is because I detected that Curator uses the resolved and cached IP address of the Zookeeper server instead of the hostname, once it is resolved. This is problematic in environment like Kubernetes, in which we see frequent change of IPs for the same pod (and `hostname`). By doing this, we force Curator to always resolve the hostname of the Zookeeper server it wants to connect to.

Reply:
Ok, went through the Curator code and observing that the `connectString` passed to `org.apache.curator.utils.ZookeeperFactory#newZooKeeper(connectString,...)` is the same as the one passed via `CuratorFrameworkFactory.builder().connectString(..)`. The dns resolution does not seem be done by the Curator client but the ZkClient. 
Let me know if you have pointers to help understand this better...